docs/proposals/kubegen.md

```diff
@@ -0,0 +1,214 @@
+# Kubernetes Code Generators
+
+Kubernetes provides a collection of code generators to assist with developing APIs.  These code generators generate
+useful libraries such as clients, openapi definitions, copying go structs, etc.
+
+The `kubegen` uses directory structure conventions to detect Kubernetes APIs definitions and run the standard
+set of Kubernetes code generators.
+
+There are 2 methods for running kubegen
+
+- Directly through the kubegen command line by downloading the binary and running at a the project root
+- Through Bazel by adding rules to `WORKSPACE` and `BUILD.gazel`
+
+## Kubernetes API conventions
+
+kubegen will run the full set of Kubernetes code generators for all APIs it finds in the project.
+kubegen looks for API groups defined under `pkg/apis/<group>/<version>` where *group* and *version* match the following
+patterns.
+
+- **group pattern**: ^[a-z]+$
+- **version pattern**: ^v\\d+(alpha\\d+|beta\\d+)*$
+
+By default, kubegen will run code generators for both external types defined under `pkg/apis/<group>/<version>` and
+internal types defined under `pkg/apis/<group>`.  The location kubegen searches can be overridden.
+
+kubegen will prepend all generated files with copyright and license comments.
+
+By default, the kubegen binary runs all* of the Kubernetes code
+generators against all API group/versions defined in the project.
+
+**List of generators:*
+
+- client-gen
+- conversion-gen
+- deepcopy-gen
+- defaulter-gen
+- go-to-protobuf
+- informer-gen
+- lister-gen
+- openapi-gen
+- set-gen
+
+## Running via Command line
+
+### Installing the command line
+
+- Download the latest kubegen release from the code-generators release page.
+- Extract the tar and add the `kubegen` binary to the PATH
+- Run `kubegen version` to list the version
+  - Will correspond to a Kubernetes release
+
+### Running the command line
+
+#### Use defaults
+
+```sh
+kubegen
+```
+
+- run all code generators against discovered APIs
+  - run for both internal and external types
+- prepend license to generated files
+  - wrap the contents of the LICENSE in comments
+  - exit non-zero if the LICENSE file is missing
+
+#### Specifying copyright owners
+
+The copyright owners for the license can be defined with the `--copyright` flag
+
+```sh
+kubegen --copyright "The Kubernetes Authors."
+```
+
+- run all code generators against discovered APIs
+  - run for both internal and external types
+- prepend license to generated files
+  - wrap the contents of the LICENSE in comments
+  - start license with "Copyright <current year> <copyright>" e.g. *Copyright 2018 The Kubernetes Authors.*
+  - exit non-zero if the LICENSE file is missing
+
+
+#### Specifying the license file
+
+The license file can be overridden from the LICENSE file using the `--license-file` flag
+
+```sh
+kubegen --license-file "boilerplate.go.txt"
```

Comment:
can we re-use the gengo flag name `--go-header-file`?

Reply:
we could, but I find this to be more descriptive and that it models the intent better.  the go-header-file is in practice just the copyright and license.  I would expect most new repos / projects to use the LICENSE file in the project root if it exists.  this flag name more closely matches the default.

Reply:
Took a moment to understand that you want something else. The "boilerplate.go.txt" example flag is wrong.

There is no reason not support both. But I like the idea to use the LICENSE file by default.

Reply:
How about we start with just `--license-file` with the description saying it specifies boilerplate.go.txt.  Having 2 flags for the same thing is a bit confusing.